File: solutions/fsharp/raindrops/1/Raindrops.fs
module Raindrops

type Sound = int * string

let sounds: Sound array = [| (3, "Pling"); (5, "Plang"); (7, "Plong") |]

let convert (n: int) =
    sounds
    |> Seq.fold
        (fun acc (f, s) ->
            match n % f with
            | 0 ->
                match acc with
                | Some(a) -> Some(a + s)
                | None -> Some(s)
            | _ -> acc)
        None
    |> fun res ->
        match res with
        | Some(s) -> s
        | None -> $"{n}"

File: solutions/elixir/raindrops/1/lib/raindrops.ex
defmodule Raindrops do
  def convert(n) do
    sound(n, [{3, "Pling"}, {5, "Plang"}, {7, "Plong"}])
  end

  defp sound(n, sounds, acc \\ "")

  defp sound(n, [{f, s} | r], acc) do
    if rem(n, f) == 0, do: sound(n, r, acc <> s), else: sound(n, r, acc)
  end

  defp sound(n, [], acc) do
    if acc != "", do: acc, else: "#{n}"
  end
end
